File: python/templates/python/app/greeter.py
import restate

from datetime import timedelta
from pydantic import BaseModel

from .utils import send_notification, send_reminder


# You can also just use a typed dict, without Pydantic
class GreetingRequest(BaseModel):
    name: str


class Greeting(BaseModel):
    message: str


greeter = restate.Service("Greeter")


@greeter.handler()
async def greet(ctx: restate.Context, req: GreetingRequest) -> Greeting:
    # Durably execute a set of steps; resilient against failures
    greeting_id = str(ctx.uuid())
    await ctx.run_typed("notification", send_notification, greeting_id=greeting_id, name=req.name)
    await ctx.sleep(timedelta(seconds=1))
    await ctx.run_typed("reminder", send_reminder, greeting_id=greeting_id, name=req.name)

    # Respond to caller
    return Greeting(message=f"You said hi to {req.name}!")

File: go/templates/go/greeter.go
package main

import (
	"time"

	restate "github.com/restatedev/sdk-go"
)

// Greeter is a struct which represents a Restate service; reflection will turn exported methods into service handlers
type Greeter struct{}

func (Greeter) Greet(ctx restate.Context, name string) (string, error) {
	//  Durably execute a set of steps; resilient against failures
	greetingId := restate.UUID(ctx).String()

	if _, err := restate.Run(ctx,
		func(ctx restate.RunContext) (restate.Void, error) {
			return SendNotification(greetingId, name)
		},
		restate.WithName("Notification"),
	); err != nil {
		return "", err
	}

	if err := restate.Sleep(ctx, 1*time.Second); err != nil {
		return "", err
	}

	if _, err := restate.Run(ctx,
		func(ctx restate.RunContext) (restate.Void, error) {
			return SendReminder(greetingId, name)
		},
		restate.WithName("Reminder"),
	); err != nil {
		return "", err
	}

	// Respond to caller
	return "You said hi to " + name + "!", nil
}
